lib/Dialect/AIEVec/Transforms/ConvertVectorToAIEVec.cpp

```diff
@@ -862,6 +863,156 @@ using LowerVectorSubIOpToAIEVecSubOp =
+                            Value rval, VectorType srcType, SrcOpTy srcOp) {
+  auto lCastOp = rewriter.create<aievec::CastOp>(srcOp.getLoc(), srcType, lval,
+                                                 /*isResAcc*/ true);
+  auto rCastOp = rewriter.create<aievec::CastOp>(srcOp.getLoc(), srcType, rval,
+                                                 /*isResAcc*/ true);
+  auto elemOp = rewriter.create<AIEv2ElemOp>(
+      srcOp.getLoc(), lCastOp->getResult(0).getType(), lCastOp->getResult(0),
+      rCastOp->getResult(0));
+  rewriter.replaceOpWithNewOp<aievec::CastOp>(
+      srcOp, srcOp.getType(), elemOp.getResult(), /*isResAcc*/ false);
+}
+
+template <typename SrcOpTy, typename DstOpTy>
+struct LowerVectorAddOrSubOpToAIEVecAddElemOrSubElemOp
+    : public OpConversionPattern<SrcOpTy> {
+  using OpConversionPattern<SrcOpTy>::OpConversionPattern;
+  using OpAdaptor = typename SrcOpTy::Adaptor;
+
+  LowerVectorAddOrSubOpToAIEVecAddElemOrSubElemOp(MLIRContext *context,
+                                                  unsigned shiftParam = 0)
+      : OpConversionPattern<SrcOpTy>(context), shiftParam(shiftParam) {}
+
+  LogicalResult
+  matchAndRewrite(SrcOpTy srcOp, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    VectorType resultType = dyn_cast<VectorType>(srcOp.getType());
+    if (!resultType)
+      return failure();
+
+    // A set recording the vector lane size and element width we are supporting
+    // for aie-ml.
+    llvm::SmallSet<std::pair<unsigned, signed>, 16> laneSizeElWidthPairSet;
+    laneSizeElWidthPairSet.insert({64, 8});
+    laneSizeElWidthPairSet.insert({32, 16});
+    laneSizeElWidthPairSet.insert({16, 32});
+    laneSizeElWidthPairSet.insert({32, 32});
+
+    auto lhs = adaptor.getLhs();
+    auto rhs = adaptor.getRhs();
+    auto lhsDefOp = lhs.getDefiningOp();
+    auto rhsDefOp = rhs.getDefiningOp();
+    if ((lhsDefOp && isa<arith::MulIOp>(lhsDefOp)) ||
+        (rhsDefOp && isa<arith::MulIOp>(rhsDefOp)) ||
+        (lhsDefOp && isa<arith::MulFOp>(lhsDefOp)) ||
+        (rhsDefOp && isa<arith::MulFOp>(rhsDefOp)))
+      return failure();
+
+    Type scalarType = resultType.getElementType();
+    unsigned resultElWidth = scalarType.getIntOrFloatBitWidth();
+    unsigned laneSize = getVectorLaneSize(resultType);
+
+    if (scalarType.isa<IntegerType>()) {
+      if (!laneSizeElWidthPairSet.count(
+              std::make_pair(laneSize, resultElWidth)))
+        return failure();
+
+      if (laneSize == 32 && resultElWidth == 32) {
+        if (!lhsDefOp || !rhsDefOp) {
+          genAddElemAieML<SrcOpTy, DstOpTy>(rewriter, lhs, rhs, resultType,
+                                            srcOp);
+          return success();
+        }
+        auto lhsExt = dyn_cast<arith::ExtSIOp>(lhsDefOp);
+        auto rhsExt = dyn_cast<arith::ExtSIOp>(rhsDefOp);
+        if (!lhsExt || !rhsExt) {
```

Comment:
Since the 'then' branch always returns, there is an opportunity to reduce the nesting level of the 'else' part by 
`if (blah) {.. ; return succss();} else { ELSE_PART }`
->
`if (blah) {.. ; return succss();} 
  ELSE_PART`
(irgnore messed up formatting)